File: algorithms/alphaCore.R
#' alphaCore ranking of nodes in a complex, directed network
#'
#' \code{alphaCore} returns a node ranking of a graph
#'
#' Iteratively computes a node ranking based on a feature set derived from
#' edge attributes and optionally static node features using the
#' mahalanobis data depth function at the origin.
#'
#' @param input_graph An igraph object of a directed graph
#' @param features A list of selected node features; default is ["all"]
#' @param stepSize defines the stepsize of each iteration as percentage of node count
#'   Higher values (e.g., > 0.1) speed up execution but yield lower ranking resolution, while lower values (e.g., < 0.1)
#'   provide finer ranking but increase runtime.
#' @param startEpsilon the epsilon to start with. Removes all nodes with depth>epsilon at start
#'   Higher values (e.g., > 1.0) remove more nodes early, emphasizing denser cores,
#'   whereas lower values (e.g., < 1.0) allow for a more gradual refinement of the ranking.
#' @param exponentialDecay dynamically reduces the step size, to have high cores with few nodes
#'   which results in higher resolution for the highest cores. Set to TRUE if node ranking precision is more important than speed.
#' @return A dataframe of node name and alpha value indicating the ranking
alphaCore <- function(input_graph,
                      features = "all",
                      stepSize = 0.1,
                      startEpsilon = 1,
                      exponentialDecay = TRUE) {
  #1 Extract numerical node features from the graph
  node_features <- extractFeatures(input_graph, features)

  #2 Compute covariance matrix
  cov_mat  <- cov(node_features[, -1, with = FALSE])
  cov_mat_inv <- tryCatch(
    solve(cov_mat),
    error = function(e1) {
      message("Covariance singular; falling back to pseudo-inverse.")
      mat <- MASS::ginv(cov_mat)
      if (nrow(mat) == 0 || ncol(mat) == 0) {
        stop("Could not invert covariance: no valid features remain.")
      }
      mat
    }
  )
  #3
  node_features$depth <- mhdOrigin(node_features[, -1, with = FALSE], cov_mat_inv)

  #4
  epsilon <- startEpsilon
  #5
  result <- node_features[, c("node")]
  result$alpha <- 0
  #6
  result$batch <- 0
  #7
  alpha <- 1 - epsilon # modified
  #8
  alpha_prev <- alpha # modified
  #9
  batch_ID <- 0
  #10
  while (vcount(input_graph) > 0) {
    #11
    repeat{
      #12: for each depth geq epsilon
      #13
      nodes <- node_features[depth >= epsilon]$node
      result[node %in% nodes, alpha := alpha_prev]
      #14
      result[node %in% nodes, batch := batch_ID]
      input_graph <- delete_vertices(input_graph, nodes)
      #16 
      batch_ID <- batch_ID + 1
      #17
      node_features <- extractFeatures(input_graph, features)
      #18
      node_features$depth <- mhdOrigin(node_features[, -1, with = FALSE], cov_mat_inv)
      
      if(length(nodes) == 0){ # 19 : if there doesn't exist a vertex with depth geq epsilon
        break
      }
    }
    #20
    alpha_prev <- alpha
    #22 # reduce epsilon with strategy
    if(exponentialDecay) {
      localStepSize <- ceiling(vcount(input_graph) * stepSize)
      epsilon <- min(head(node_features[order(depth, decreasing = T)], localStepSize)$depth)
    } else {
      epsilon <- epsilon - stepSize
    }
    #21
    alpha <- 1 - epsilon
    
  }
  return(result)
  #
}


############################## AUX FUNCTIONS ################################

#' Compute default node features for alphaCore
#'
#' \code{computeNodeFeaturFun} returns a data.table with indegree and strength
#'
#' @param graph An igraph object
#' @return A data.table with node names, indegree and strength
computeNodeFeaturFun <- function(graph) {

  # get node names
  nodes <- V(graph)$name
  # compute indegree
  indegree <- degree(graph, mode = "in")
  # compute sum of incoming weights (a.k.a. "strength")
  strength <- strength(graph, mode="in")
  # combine results
  nodeFeatures <- data.table(node=nodes, indegree=indegree, strength=strength)
  return(nodeFeatures)
}


#' Compute specified node features
#'
#' \code{computeNodeFeatures} calculates requested network features for each node
#'
#' @param graph An igraph object
#' @param features A character vector of features to compute
#' @return A data.table with node names and the requested features
computeNodeFeatures <- function(graph, features = c("indegree","strength")) {
  nodeFeatures <- data.table(node = V(graph)$name)
  if("degree" %in% features) {
    nodeFeatures[, indegree := degree(graph, mode = "all")]
  }
  if("indegree" %in% features) {
    nodeFeatures[, indegree := degree(graph, mode = "in")]
  }
  if("outdegree" %in% features) {
    nodeFeatures[, outdegree := degree(graph, mode = "out")]
  }
  if("strength" %in% features) {
    nodeFeatures[, strength := strength(graph, mode = "all")]
  }
  if("instrength" %in% features) {
    nodeFeatures[, instrength := strength(graph, mode = "in")]
  }
  if("outstrength" %in% features) {
    nodeFeatures[, outstrength := strength(graph, mode = "out")]
  }
  if("triangles" %in% features) {
    nodeFeatures[, triangles := count_triangles(graph)]
  }
  if("neighborhoodsize" %in% features) {
    nodeFeatures[, neighborhoodsize := neighborhood.size(graph, mode = "all", mindist = 1)]
  }
  if("inneighborhoodsize" %in% features) {
    nodeFeatures[, inneighborhoodsize := neighborhood.size(graph, mode = "in", mindist = 1)]
  }
  if("outneighborhoodsize" %in% features) {
    nodeFeatures[, outneighborhoodsize := neighborhood.size(graph, mode = "out", mindist = 1)]
  }
  return(nodeFeatures[])
}

customNodeFeatures <- function(features) {
  return(function(x) {
    return(computeNodeFeatures(x, features = features))
  })
}


#' Extract numerical node features from a graph
#'
#' \code{extractFeatures} returns a data.table of node features
#'
#' Extracts numerical vertex attributes from the graph. If specified features
#' don't exist as vertex attributes, they will be computed on demand.
#'
#' @param graph An igraph object
#' @param features Features to extract: "all" for all numeric attributes, or a character vector of specific features
#' @return A data.table with node names and requested features
extractFeatures <- function(graph, features = "all") {
  #If the user explicitly requested a feature list, compute those
  if (!identical(features, "all")) {
    df <- computeNodeFeatures(graph, features = features)
    missing <- setdiff(features, names(df))
    if (length(missing) > 0L) {
      stop(
        "Vertex attribute(s) not found or not numeric: ",
        paste(missing, collapse = ", "),
        call. = FALSE
      )
    }
    return(df)
  }

  #Otherwise, pull *all* numeric vertex attributes off the graph
  attrs <- vertex.attributes(graph)
  numeric_names <- names(attrs)[sapply(attrs, is.numeric)]

  #If none are present, fall back to the original indegree+strength
  if (length(numeric_names) == 0L) {
    message(
      "No numeric vertex attributes found; reverting to indegree+strength."
    )
    return(computeNodeFeaturFun(graph))
  }

  #Build a data.table: one column 'node', then each numeric attribute
  dt <- data.table(node = V(graph)$name)
  for (nm in numeric_names) dt[[nm]] <- attrs[[nm]]
  return(dt[])
}



#' Calculate the Mahalanobis depth to the origin
#'
#' \code{mhdOrigin} calculates the Mahalanobis depth of data points to the origin
#'
#' @param data A matrix or data.frame of multivariate data
#' @param sigma_inv The inverse covariance matrix
#' @return A vector of depth values
mhdOrigin <- function(data, sigma_inv) {
  origin <- rep(0,ncol(data)) # c(0,0,...)
  # We reuse the Mahalanobis distance implementation of the stats package,
  # which returns the squared Mahalanobis distance: (x - μ)' Σ^-1 (x - μ) = D^2
  # To arrive at the Mahalanobis Depth to the origin, we only need to add 1 and
  # take the reciprocal.
  return((1 + pmax(stats::mahalanobis(data, center = origin, sigma_inv, inverted = TRUE), 0))^-1)
}

File: algorithms/alphacore.py
import networkx as nx
import numpy as np
import pandas as pd
import math

# alphaCore ranking of nodes in a complex, directed network
#
# Iteratively computes a node ranking based on a feature set derived from
# edge attributes and optionally static node features using the
# mahalanobis data depth function at the origin.
#
# @param graph A networkx directed graph.
# @param features A list of selected node features; default is ["all"]
# @param stepSize Defines the stepsize of each iteration as percentage of node count
#   Higher values (>0.1) speed up execution but lower ranking resolution.
#   Lower values (<0.1) provide finer ranking but increase runtime.
# @param startEpsi The epsilon to start with. Removes all nodes with depth>epsilon at start
#   Higher values (>1.0) remove more nodes early, emphasizing denser cores.
#   Lower values (<1.0) refine ranking progressively.
# @param expoDecay Dynamically reduces the step size, to have high cores with few nodes if true
#   Recommended when ranking precision is more important than speed.
# @return A dataframe of columns nodeID, alpha value, and batchID

def alphaCore(graph, features=["all"], stepSize = 0.1, startEpsi = 1, expoDecay = False):
    #1 Extract numerical node features from the graph; if unavailable, use computeNodeFeatures.
    data = extractFeatures(graph, features)

    #2 compute cov matrix to be used for all remainder of depth calculations
    matrix = data.drop("nodeID", axis=1)  # convert dataframe to numeric matrix by removing first column containing nodeID
    cov = np.cov(matrix.values.T)
    
    #3 calculate the Mahalanobis depth and add it to the respective row of the dataframe
    data['mahal'] = calculateMahalFromCenter(data, 0, cov)
    
    #4 
    epsi = startEpsi
    #5
    node = []
    alphaVals = []
    #6
    batch = []
    #7
    alpha = 1 - epsi
    #8
    alphaPrev = alpha
    #9
    batchID = 0
    
    #10 
    while graph.number_of_nodes() > 0:
        #11
        while True:
            depthFound = False  # to simulate do-while loop; used to check if there exists a node with depth >= epsi on current iteration
            #12 
            for row in data.itertuples():
                if row.mahal >= epsi:
                    depthFound = True
                    #13
                    node.append(row.nodeID)  # set node core
                    alphaVals.append(alphaPrev)
                    #14
                    batch.append(batchID)
                    #15
                    graph.remove_node(row.nodeID)
            #16
            batchID += 1
            #19 while condition of do-while loop of #11
            if graph.number_of_nodes() == 0 or not depthFound:
                break
            #17
            data = extractFeatures(graph, features)  # recompute node properties
            #18
            data['mahal'] = calculateMahalFromCenter(data, 0, cov)  # recompute depth
        #20 
        alphaPrev = alpha
        #21 
        if expoDecay and graph.number_of_nodes() > 0:  # exponential decay
            localStepSize = math.ceil(graph.number_of_nodes() * stepSize)
            data = data.sort_values(ascending=False, by=['mahal'])
            epsi = data.iloc[localStepSize - 1]['mahal']
        else:  # step decay
            epsi -= stepSize
        #22
        alpha = 1 - epsi
    #23
    return pd.DataFrame({'nodeID': node, 'alpha': alphaVals, 'batchID': batch})


# Extract numerical node features from the graph.
# 
# @param graph A NetworkX directed graph.
# @param features A list of feature names to extract. If None, defaults to ["all"].
# @return A DataFrame containing extracted numerical node features.
def extractFeatures(graph, features=["all"]):
    #Collect all node attributes and identify numerical features
    all_features = set()
    numeric_features = set()
    selected_features = set()
    data = {}

    for node_id, attributes in graph.nodes(data=True):
        for key, value in attributes.items():
            all_features.add(key)  # Track all available features
            if isinstance(value, (int, float)):  
                numeric_features.add(key)  # Track only numeric features

    #Determine the features to extract
    if features == ["all"]:
        if not numeric_features:
            print("No numerical node features found. Reverting to default AlphaCore node features.")
            return computeNodeFeatures(graph)
        selected_features = numeric_features  # Use all numeric features
    else:
        for feature in features:
            if feature not in all_features:
                print(f"Debug: Available features: {all_features}")
                raise ValueError(f"Feature '{feature}' not found in graph nodes. Please check your graph.")
            if feature in numeric_features:
                selected_features.add(feature)

        if not selected_features:
            print("None of the selected features contain numerical values. Reverting to default AlphaCore node features.")
            return computeNodeFeatures(graph)


    #Extract feature values for each node
    for node_id, attributes in graph.nodes(data=True):
        node_features = {}
        for feature in selected_features:
            if feature in attributes:
                node_features[feature] = attributes.get(feature)  
        data[node_id] = node_features

    #Convert to DataFrame
    df = pd.DataFrame.from_dict(data, orient="index").reset_index().rename(columns={"index": "nodeID"})
    return df



# Computes the node features of a given directed graph and returns a dataframe containing the features of each node
#
# @param graph A networkx directed graph
# @return A dataframe containing the computed node features with each row as a new entry and columns as different features
def computeNodeFeatures(graph):

    # additional or different node features can be computed and added as a column to the dataframe in the same manner as below

    nodeID = []
    inDegree = []
    outDegree = []
    inStrength = []
    outStrength = []
    for node in graph:
        nodeID.append(node)
        inDegree.append(graph.in_degree(node))
        outDegree.append(graph.out_degree(node))
        inStrength.append(graph.in_degree(node, "value"))
        outStrength.append(graph.out_degree(node, "value"))

    # currently adding inDegree, outDegree, inStrength, and outStrength to dataframe
    df = pd.DataFrame({"nodeID": nodeID, "inDegree": inDegree, "inStrength": inStrength, "outDegree": outDegree,
                       "outStrength": outStrength})
    return df



# Computes the mahalanobis depth from a given center of each row of a given set of data and returns it as an array
#
# @param data Dataframe where each row is a new entry and each column after the first (nodeID) is a type of data
# @param center A center value calculated with respect to when computing mahalanobis depth
# @param cov The covariance matrix of the data matrix
# @return An array containing the mahalanobis depth of each row entry of a given set of data
def calculateMahalFromCenter(data, center, cov):
    matrix = data.drop("nodeID", axis=1)  # convert dataframe to numeric matrix by removing first column containing nodeID
    x_minus_center = matrix.values - center
    x_minus_center_transposed = x_minus_center.T
    # Try computing the inverse of the covariance matrix; if it fails, fall back to the pseudo-inverse for stability.
    try: 
        inv_cov = np.linalg.inv(cov)
    except np.linalg.LinAlgError:
        print("Covariance matrix is not invertible, using Moore-Penrose pseudo-inverse instead.")  
        inv_cov = np.linalg.pinv(cov)
    left = np.dot(x_minus_center, inv_cov)
    mahal = np.dot(left, x_minus_center_transposed)
    mahal_diag = np.diagonal(mahal)   # diagonal contains the depth vaulues corresponding to each row from matrix
    return np.reciprocal(1 + np.maximum(mahal_diag, 0))  # Ensure stability
